component/azstorage/block_blob.go

```diff
@@ -653,6 +655,195 @@ func (bb *BlockBlob) WriteFromBuffer(name string, metadata map[string]string, da
+	}
+	return &blockList, nil
+}
+
+// create our definition of block
+func (bb *BlockBlob) createBlock(blockIdLength, startIndex, size int64) *common.Block {
+	newBlockId := base64.StdEncoding.EncodeToString(common.NewUUID(blockIdLength))
+	newBlock := &common.Block{
+		Id:         newBlockId,
+		StartIndex: startIndex,
+		EndIndex:   startIndex + size,
+		Size:       size,
+	}
+	return newBlock
+}
+
+func (bb *BlockBlob) createNewBlocks(modBlockList, blockList *common.BlockOffsetList, offset, length, blockIdLength int64) (bool, int64) {
+	prevIndex := blockList.BlockOffsetList[len(blockList.BlockOffsetList)-1].EndIndex
+	// BufferSize is the size of the buffer that will go beyond our current blob (appended)
+	var bufferSize int64
+	// counter will help us keep track of how many blocks we've created
+	counter := int64(0)
+	// appendOnly means there is no overlap at all with the blob - data only being appended
+	appendOnly := false
+	// if modBlockList is not of size 0 then we know that we're overwriting part of our current data - with new data also being appended to the end
+	if len(modBlockList.BlockOffsetList) != 0 {
+		bufferSize = length - (prevIndex - offset)
+		// only new data at the end basically offset => lastBlock.EndIndex
+	} else {
+		bufferSize = (offset + length) - prevIndex
+		appendOnly = true
+	}
+	// reset the offset to be the last index in the block list since we are adding new blocks
+	offset = prevIndex
+	startIndex := offset
+	for i := 0; i < int(bufferSize); i++ {
+		counter += 1
+		// create a new block if we hit our block size
+		if int64(i)%bb.Config.blockSize == 0 && i != 0 {
+			newBlock := bb.createBlock(blockIdLength, startIndex, bb.Config.blockSize)
+			modBlockList.BlockOffsetList = append(modBlockList.BlockOffsetList, newBlock)
+			blockList.BlockOffsetList = append(blockList.BlockOffsetList, newBlock)
+			startIndex = newBlock.EndIndex
+			// reset the counter since it will help us to determine if there is leftovers at the end
+			counter = 0
+		}
+	}
+	if counter != 0 {
+		newBlock := bb.createBlock(blockIdLength, startIndex, counter)
+		modBlockList.BlockOffsetList = append(modBlockList.BlockOffsetList, newBlock)
+		blockList.BlockOffsetList = append(blockList.BlockOffsetList, newBlock)
+	}
+	return appendOnly, bufferSize
+}
+
+// Write : write data at given offset to a blob
+func (bb *BlockBlob) Write(name string, offset, length int64, data []byte, fileOffsets, modFileOffsets *common.BlockOffsetList) error {
+	defer log.TimeTrack(time.Now(), "BlockBlob::Write", name)
+	log.Trace("BlockBlob::Write : name %s offset %v", name, offset)
+	blobURL := bb.Container.NewBlockBlobURL(filepath.Join(bb.Config.prefixPath, name))
+	// tracks the case where our offset is great than our current file size (appending only - not modifying pre-existing data)
+	appendOnly := false
+	var dataBuffer *[]byte
+
+	// when the file offset mapping is cached we don't need to make a get block list call
+	if !fileOffsets.Cached {
```

Comment:
fileOffsets and modFileOffsets are of pointer type here and we shall check them against nil for safety.
As these are not populated in fuse2, when I tried running this code it was crashing.